components/lcd/esp_lcd_st7796/esp_lcd_st7796.c

```diff
@@ -1,353 +1,38 @@
-        io_conf.mode = GPIO_MODE_OUTPUT;
-        io_conf.pin_bit_mask = 1ULL << panel_dev_config->reset_gpio_num;
-        ESP_GOTO_ON_ERROR(gpio_config(&io_conf), err, TAG, "configure GPIO for RST line failed");
-    }
+    esp_err_t ret = ESP_ERR_NOT_SUPPORTED;
 
-#if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(5, 0, 0)
-    switch (panel_dev_config->color_space) {
-    case ESP_LCD_COLOR_SPACE_RGB:
-        st7796->madctl_val = 0;
-        break;
-    case ESP_LCD_COLOR_SPACE_BGR:
-        st7796->madctl_val |= LCD_CMD_BGR_BIT;
-        break;
-    default:
-        ESP_GOTO_ON_FALSE(false, ESP_ERR_NOT_SUPPORTED, err, TAG, "unsupported color space");
-        break;
-    }
-#else
-    switch (panel_dev_config->rgb_endian) {
-    case LCD_RGB_ENDIAN_RGB:
-        st7796->madctl_val = 0;
-        break;
-    case LCD_RGB_ENDIAN_BGR:
-        st7796->madctl_val |= LCD_CMD_BGR_BIT;
-        break;
-    default:
-        ESP_GOTO_ON_FALSE(false, ESP_ERR_NOT_SUPPORTED, err, TAG, "unsupported rgb endian");
-        break;
+#if SOC_MIPI_DSI_SUPPORTED
+    if (vendor_config->flags.use_mipi_interface) {
+        ret = esp_lcd_new_panel_st7796_mipi(io, panel_dev_config, ret_panel);
+        return ret;
     }
 #endif
 
-    switch (panel_dev_config->bits_per_pixel) {
-    case 16: // RGB565
-        st7796->colmod_val = 0x05;
-        st7796->fb_bits_per_pixel = 16;
-        break;
-    case 18: // RGB666
-        st7796->colmod_val = 0x06;
-        // each color component (R/G/B) should occupy the 6 high bits of a byte, which means 3 full bytes are required for a pixel
-        st7796->fb_bits_per_pixel = 24;
-        break;
-    case 24: // RGB888
-        st7796->colmod_val = 0x07;
-        st7796->fb_bits_per_pixel = 24;
-        break;
-    default:
-        ESP_GOTO_ON_FALSE(false, ESP_ERR_NOT_SUPPORTED, err, TAG, "unsupported pixel width");
-        break;
-    }
-
-    st7796->io = io;
-    st7796->reset_gpio_num = panel_dev_config->reset_gpio_num;
-    st7796->reset_level = panel_dev_config->flags.reset_active_high;
-    if (panel_dev_config->vendor_config) {
-        st7796->init_cmds = ((st7796_vendor_config_t *)panel_dev_config->vendor_config)->init_cmds;
-        st7796->init_cmds_size = ((st7796_vendor_config_t *)panel_dev_config->vendor_config)->init_cmds_size;
-    }
-    st7796->base.del = panel_st7796_del;
-    st7796->base.reset = panel_st7796_reset;
-    st7796->base.init = panel_st7796_init;
-    st7796->base.draw_bitmap = panel_st7796_draw_bitmap;
-    st7796->base.invert_color = panel_st7796_invert_color;
-    st7796->base.set_gap = panel_st7796_set_gap;
-    st7796->base.mirror = panel_st7796_mirror;
-    st7796->base.swap_xy = panel_st7796_swap_xy;
-#if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(5, 0, 0)
-    st7796->base.disp_off = panel_st7796_disp_on_off;
-#else
-    st7796->base.disp_on_off = panel_st7796_disp_on_off;
+#if SOC_LCD_I80_SUPPORTED
+    ret = esp_lcd_new_panel_st7796_general(io, panel_dev_config, ret_panel);
```

Comment:
When chip supports I80 and MIPI interface and user select `use_mipi_interface`, should be called this function? 

And what about SPI usage?

Reply:
This was an oversight; I forgot that the ST7796 also supports SPI. In fact, my intention was to use the `esp_lcd_new_panel_st7796_mipi` interface when the `use_mipi_interface` bit is enabled. When using `vendor_config`, or when `vendor_config` is used but `use_mipi_interface` is not enabled, the default should be the `esp_lcd_new_panel_st7796_general` interface.